# (PART) Analysis steps {-}

# Analysis steps

This part consists of several chapters for steps in a computational analysis pipeline for spatially resolved transcriptomics (ST) data. This includes quality control (QC), normalization, feature selection, dimensionality reduction, clustering, and identifying marker genes.

These steps require that the raw data has been loaded into R. In the previous part, we provide instructions and examples showing how to do this for the 10x Genomics Visium platform.

Throughout these chapters, we follow the Bioconductor principle of modularity -- for steps where multiple alternative analysis tools exist, we show how to substitute these, using the `SpatialExperiment` class to store input and output objects.


## Load data

In the following analysis chapters, we use a pre-prepared dataset where we have previously applied the steps described in [Preprocessing steps], and saved the object in the `SpatialExperiment` format. This is available from the [STexampleData](https://github.com/lmweber/STexampleData) package.

The dataset consists of a single sample of human brain from the dorsolateral prefrontal cortex (DLPFC) region, measured using the 10x Genomics Visium platform, from our publication @Maynard2021. The dataset is described in more detail in [Visium human DLPFC workflow].

Here, we show how to load the data from the [STexampleData](https://github.com/lmweber/STexampleData) package.

```{r load_data, message=FALSE}
library(SpatialExperiment)
library(STexampleData)

# load object
spe <- Visium_humanDLPFC()
```


## SpatialExperiment object structure

Next, we inspect the `SpatialExperiment` object. For more details, see [SpatialExperiment].

```{r, message=FALSE}
# check object
spe

# number of features (rows) and spots (columns)
dim(spe)

# names of 'assay' tables
assayNames(spe)

# features metadata
head(rowData(spe))

# spots metadata
head(colData(spe))

# spatial metadata
head(spatialData(spe))

# spatial coordinates
head(spatialCoords(spe))

# image metadata
imgData(spe)
```

