# Load data {#sec-seq-load-data}

## Introduction

The following chapters provide examples demonstrating methods for individual analysis steps for spatial transcriptomics data from sequencing-based platforms.

In these chapters, we assume the datasets are formatted as `SpatialExperiment` objects (see @sec-bkg-bioconductor-data-classes).

Here, we load a 10x Genomics Visium dataset that will be used in several of the following chapters. This dataset has previously been preprocessed using data preprocessing procedures with tools outside R and saved in `SpatialExperiment` format. (For more details on data preprocessing procedures for the Visium platform, see the related online book [Visium Data Preprocessing](https://lmweber.org/Visium-data-preprocessing/), also listed in @sec-apx-related-resources-data-preprocessing.) This dataset is available for download in `SpatialExperiment` format from the [STexampleData](https://bioconductor.org/packages/STexampleData) Bioconductor package.


## Dataset

This dataset consists of one sample (Visium capture area) from one donor, consisting of postmortem human brain tissue from the dorsolateral prefrontal cortex (DLPFC) brain region, measured with the 10x Genomics Visium platform. The dataset is described in @Maynard2021.

More details on the dataset are also included in @sec-seq-workflow-human-dlpfc.

<!-- To do: additional details on dataset here (from DLPFC workflow chapter) -->


## Load data

Download and load the dataset in `SpatialExperiment` format from the [STexampleData](https://bioconductor.org/packages/STexampleData) Bioconductor package.

```{r, message=FALSE}
library(SpatialExperiment)
library(STexampleData)
```

```{r}
# load object
spe <- Visium_humanDLPFC()
# temporary fix
imgData(spe) <- NULL
```


## Save objects for later chapters {#sec-seq-load-data-save-objects}

We also save the object(s) in `.rds` format for re-use within later chapters to speed up the build time of the book.

```{r}
# save object(s)
saveRDS(spe, file = "spe_load.rds")
```


## SpatialExperiment object

Check the structure of the `SpatialExperiment` object. For more details on the `SpatialExperiment` structure, see @sec-bkg-bioconductor-data-classes.

```{r}
# check object
spe

# number of genes (rows) and spots (columns)
dim(spe)

# names of 'assays'
assayNames(spe)

# row (gene) data
head(rowData(spe))

# column (spot) data
head(colData(spe))

# spatial coordinates
head(spatialCoords(spe))

# image data
imgData(spe)
```


## Build object

Alternatively, we can also build a `SpatialExperiment` object directly from raw data.

Here, we provide a short example with an empty dataset.

For more details, including how to load raw data from the 10x Genomics Space Ranger output files to build an object, or how to add image data to the object, see the [SpatialExperiment](https://bioconductor.org/packages/SpatialExperiment) documentation.

```{r}
# create data
n_genes <- 200
n_spots <- 100

counts <- matrix(0, nrow = n_genes, ncol = n_spots)

row_data <- DataFrame(
  gene_name = paste0("gene", sprintf("%03d", seq_len(n_genes)))
)

col_data <- DataFrame(
  sample_id = rep("sample01", n_spots)
)

spatial_coords <- matrix(0, nrow = n_spots, ncol = 2)
colnames(spatial_coords) <- c("x", "y")

# create SpatialExperiment object
spe <- SpatialExperiment(
  assays = list(counts = counts), 
  colData = col_data, 
  rowData = row_data, 
  spatialCoords = spatial_coords
)
```


## References {.unnumbered}
