<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Preamble â€“ Orchestrating Spatial Transcriptomics Analysis with Bioconductor</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/bkg-introduction.html" rel="next">
<link href="../index.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/preamble.html">Preamble</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">OSTA</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/bkg-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/bkg-spatial-transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Spatial transcriptomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/bkg-bioconductor-data-classes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Bioconductor data classes</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Sequencing-based platforms</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Overview</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-load-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Load data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-quality-control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Quality control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Normalization</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-feature-selection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Feature selection</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-dimensionality-reduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Dimensionality reduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-spot-deconvolution.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Spot deconvolution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-spatial-co-localization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Spatial co-localization</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-differential-expression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Differential expression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-code-summary-analysis-steps.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Code summary: Analysis steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-workflow-human-dlpfc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Workflow: Human DLPFC</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-workflow-mouse-coronal.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Workflow: Mouse coronal</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/seq-workflow-spatialibd.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Workflow: spatialLIBD</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Imaging-based platforms</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/img-overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Overview</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/img-load-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Load data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/img-workflow-imaging-dataset-x.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Workflow: Imaging-based dataset</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/apx-acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Acknowledgments</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/apx-related-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Related resources</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#docker-image" id="toc-docker-image" class="nav-link active" data-scroll-target="#docker-image">Docker image</a></li>
  <li><a href="#rstudio-server" id="toc-rstudio-server" class="nav-link" data-scroll-target="#rstudio-server">RStudio Server</a></li>
  <li><a href="#biocbook" id="toc-biocbook" class="nav-link" data-scroll-target="#biocbook">BiocBook</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Preamble</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p><strong>Package:</strong> OSTA<br><strong>Compiled:</strong> 2025-02-07<br><strong>Package version:</strong> 0.99.0<br><strong>R version:</strong> R Under development (unstable) (2025-01-28 r87664)<br><strong>Bioconductor version:</strong> 3.21<br><strong>License:</strong> CC BY 4.0<br></p>
<section id="docker-image" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="docker-image">Docker image</h2>
<p>A <code>Docker</code> image built from this repository is available here: <a href="https://ghcr.io/lmweber/osta">ghcr.io/lmweber/osta</a></p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Get started now
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can get access to all the packages used in this book in &lt;1 minute, using this command in a terminal:</p>
<div class="cell" data-layout-align="center">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>bash</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> ghcr.io/lmweber/osta:devel R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="rstudio-server" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="rstudio-server">RStudio Server</h2>
<p>An RStudio Server instance can be initiated from the <code>Docker</code> image as follows:</p>
<div class="cell" data-layout-align="center">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>bash</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="dt">\</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--volume</span> <span class="op">&lt;</span>local_folder<span class="op">&gt;</span>:<span class="op">&lt;</span>destination_folder<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-e</span> PASSWORD=OHCA <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-p</span> 8787:8787 <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    ghcr.io/lmweber/osta:devel</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The initiated RStudio Server instance will be available at: <a href="https://localhost:8787">https://localhost:8787</a></p>
</section><section id="biocbook" class="level2"><h2 class="anchored" data-anchor-id="biocbook">BiocBook</h2>
<p>This book is built using the <a href="https://github.com/js2264/BiocBook">BiocBook</a> package, developed by <a href="https://jserizay.com/">Jacques Serizay</a> for building Quarto books with Bioconductor.</p>
<p>More details on how to build a book using BiocBook can be found in the <a href="https://jserizay.com/BiocBookDemo/devel/">BiocBook book</a>.</p>
</section><section id="session-info" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="session-info">Session info</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to expand
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://sessioninfo.r-lib.org/reference/session_info.html">session_info</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Package"</span><span class="op">]</span>, </span>
<span>    include_base <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="code_output"><code>##  â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
##   setting  value
##   version  R Under development (unstable) (2025-01-28 r87664)
##   os       Ubuntu 24.04.1 LTS
##   system   x86_64, linux-gnu
##   ui       X11
##   language (EN)
##   collate  C
##   ctype    en_US.UTF-8
##   tz       Etc/UTC
##   date     2025-02-07
##   pandoc   3.6.2 @ /usr/bin/ (via rmarkdown)
##   quarto   1.5.57 @ /usr/local/bin/quarto
##  
##  â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
##   package              * version   date (UTC) lib source
##   abind                  1.4-8     2024-09-12 [2] CRAN (R 4.5.0)
##   AnnotationDbi          1.69.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   AnnotationHub          3.15.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   askpass                1.2.1     2024-10-04 [2] CRAN (R 4.5.0)
##   assorthead             1.1.13    2025-01-17 [2] Bioconductor 3.21 (R 4.5.0)
##   attempt                0.3.1     2020-05-03 [2] CRAN (R 4.5.0)
##   base                 * 4.5.0     2025-01-29 [3] local
##   base64enc              0.1-3     2015-07-28 [2] CRAN (R 4.5.0)
##   beachmat               2.23.6    2024-12-22 [2] Bioconductor 3.21 (R 4.5.0)
##   beeswarm               0.4.0     2021-06-01 [2] CRAN (R 4.5.0)
##   benchmarkme            1.0.8     2022-06-12 [2] CRAN (R 4.5.0)
##   benchmarkmeData        1.0.4     2020-04-23 [2] CRAN (R 4.5.0)
##   BH                     1.87.0-1  2024-12-17 [2] CRAN (R 4.5.0)
##   Biobase                2.67.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocFileCache          2.15.1    2025-01-17 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocGenerics           0.53.6    2025-01-27 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocIO                 1.17.1    2024-11-21 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocManager            1.30.25   2024-08-28 [2] CRAN (R 4.5.0)
##   BiocNeighbors          2.1.2     2024-12-09 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocParallel           1.41.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocSingular           1.23.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocStyle              2.35.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   BiocVersion            3.21.1    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   Biostrings             2.75.3    2024-12-15 [2] Bioconductor 3.21 (R 4.5.0)
##   bit                    4.5.0.1   2024-12-03 [2] CRAN (R 4.5.0)
##   bit64                  4.6.0-1   2025-01-16 [2] CRAN (R 4.5.0)
##   bitops                 1.0-9     2024-10-03 [2] CRAN (R 4.5.0)
##   blob                   1.2.4     2023-03-17 [2] CRAN (R 4.5.0)
##   bluster                1.17.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   bookdown               0.42      2025-01-07 [2] CRAN (R 4.5.0)
##   boot                   1.3-31    2024-08-28 [3] CRAN (R 4.5.0)
##   brew                   1.0-10    2023-12-16 [2] CRAN (R 4.5.0)
##   brio                   1.1.5     2024-04-24 [2] CRAN (R 4.5.0)
##   BRISC                  1.0.6     2024-10-02 [2] CRAN (R 4.5.0)
##   bslib                  0.9.0     2025-01-30 [2] CRAN (R 4.5.0)
##   cachem                 1.1.0     2024-05-16 [2] CRAN (R 4.5.0)
##   Cairo                  1.6-2     2023-11-28 [2] CRAN (R 4.5.0)
##   callr                  3.7.6     2024-03-25 [2] CRAN (R 4.5.0)
##   circlize               0.4.16    2024-02-20 [2] CRAN (R 4.5.0)
##   class                  7.3-23    2025-01-01 [3] CRAN (R 4.5.0)
##   cli                    3.6.3     2024-06-21 [2] CRAN (R 4.5.0)
##   clipr                  0.8.0     2022-02-22 [2] CRAN (R 4.5.0)
##   clue                   0.3-66    2024-11-13 [2] CRAN (R 4.5.0)
##   cluster                2.1.8     2024-12-11 [3] CRAN (R 4.5.0)
##   codetools              0.2-20    2024-03-31 [3] CRAN (R 4.5.0)
##   colorspace             2.1-1     2024-07-26 [2] CRAN (R 4.5.0)
##   commonmark             1.9.2     2024-10-04 [2] CRAN (R 4.5.0)
##   compiler               4.5.0     2025-01-29 [3] local
##   ComplexHeatmap         2.23.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   config                 0.3.2     2023-08-30 [2] CRAN (R 4.5.0)
##   cowplot                1.1.3     2024-01-22 [2] CRAN (R 4.5.0)
##   cpp11                  0.5.1     2024-12-04 [2] CRAN (R 4.5.0)
##   crayon                 1.5.3     2024-06-20 [2] CRAN (R 4.5.0)
##   credentials            2.0.2     2024-10-04 [2] CRAN (R 4.5.0)
##   crosstalk              1.2.1     2023-11-23 [2] CRAN (R 4.5.0)
##   curl                   6.2.0     2025-01-23 [2] CRAN (R 4.5.0)
##   data.table             1.16.4    2024-12-06 [2] CRAN (R 4.5.0)
##   datasets             * 4.5.0     2025-01-29 [3] local
##   DBI                    1.2.3     2024-06-02 [2] CRAN (R 4.5.0)
##   dbplyr                 2.5.0     2024-03-19 [2] CRAN (R 4.5.0)
##   DelayedArray           0.33.5    2025-02-03 [2] Bioconductor 3.21 (R 4.5.0)
##   desc                   1.4.3     2023-12-10 [2] CRAN (R 4.5.0)
##   devtools               2.4.5     2022-10-11 [2] CRAN (R 4.5.0)
##   diffobj                0.3.5     2021-10-05 [2] CRAN (R 4.5.0)
##   digest                 0.6.37    2024-08-19 [2] CRAN (R 4.5.0)
##   docopt                 0.7.1     2020-06-24 [2] CRAN (R 4.5.0)
##   doParallel             1.0.17    2022-02-07 [2] CRAN (R 4.5.0)
##   downlit                0.4.4     2024-06-10 [2] CRAN (R 4.5.0)
##   dplyr                  1.1.4     2023-11-17 [2] CRAN (R 4.5.0)
##   dqrng                  0.4.1     2024-05-28 [2] CRAN (R 4.5.0)
##   DT                     0.33      2024-04-04 [2] CRAN (R 4.5.0)
##   edgeR                  4.5.2     2025-01-24 [2] Bioconductor 3.21 (R 4.5.0)
##   ellipsis               0.3.2     2021-04-29 [2] CRAN (R 4.5.0)
##   evaluate               1.0.3     2025-01-10 [2] CRAN (R 4.5.0)
##   ExperimentHub          2.15.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   fansi                  1.0.6     2023-12-08 [2] CRAN (R 4.5.0)
##   farver                 2.1.2     2024-05-13 [2] CRAN (R 4.5.0)
##   fastmap                1.2.0     2024-05-15 [2] CRAN (R 4.5.0)
##   filelock               1.0.3     2023-12-11 [2] CRAN (R 4.5.0)
##   FNN                    1.1.4.1   2024-09-22 [2] CRAN (R 4.5.0)
##   fontawesome            0.5.3     2024-11-16 [2] CRAN (R 4.5.0)
##   foreach                1.5.2     2022-02-02 [2] CRAN (R 4.5.0)
##   foreign                0.8-88    2025-01-12 [3] CRAN (R 4.5.0)
##   formatR                1.14      2023-01-17 [2] CRAN (R 4.5.0)
##   fs                     1.6.5     2024-10-30 [2] CRAN (R 4.5.0)
##   futile.logger          1.4.3     2016-07-10 [2] CRAN (R 4.5.0)
##   futile.options         1.0.1     2018-04-20 [2] CRAN (R 4.5.0)
##   generics               0.1.3     2022-07-05 [2] CRAN (R 4.5.0)
##   GenomeInfoDb           1.43.4    2025-01-24 [2] Bioconductor 3.21 (R 4.5.0)
##   GenomeInfoDbData       1.2.13    2025-02-07 [2] Bioconductor
##   GenomicAlignments      1.43.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   GenomicRanges          1.59.1    2024-11-15 [2] Bioconductor 3.21 (R 4.5.0)
##   gert                   2.1.4     2024-10-14 [2] CRAN (R 4.5.0)
##   GetoptLong             1.0.5     2020-12-15 [2] CRAN (R 4.5.0)
##   ggbeeswarm             0.7.2     2023-04-29 [2] CRAN (R 4.5.0)
##   ggplot2                3.5.1     2024-04-23 [2] CRAN (R 4.5.0)
##   ggrastr                1.0.2     2023-06-01 [2] CRAN (R 4.5.0)
##   ggrepel                0.9.6     2024-09-07 [2] CRAN (R 4.5.0)
##   ggside                 0.3.1     2024-03-01 [2] CRAN (R 4.5.0)
##   ggspavis               1.13.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   gh                     1.4.1     2024-03-28 [2] CRAN (R 4.5.0)
##   gitcreds               0.1.2     2022-09-08 [2] CRAN (R 4.5.0)
##   GlobalOptions          0.1.2     2020-06-10 [2] CRAN (R 4.5.0)
##   glue                   1.8.0     2024-09-30 [2] CRAN (R 4.5.0)
##   golem                  0.5.1     2024-08-27 [2] CRAN (R 4.5.0)
##   graphics             * 4.5.0     2025-01-29 [3] local
##   grDevices            * 4.5.0     2025-01-29 [3] local
##   grid                   4.5.0     2025-01-29 [3] local
##   gridExtra              2.3       2017-09-09 [2] CRAN (R 4.5.0)
##   gtable                 0.3.6     2024-10-25 [2] CRAN (R 4.5.0)
##   here                   1.0.1     2020-12-13 [2] CRAN (R 4.5.0)
##   highr                  0.11      2024-05-26 [2] CRAN (R 4.5.0)
##   htmltools              0.5.8.1   2024-04-04 [2] CRAN (R 4.5.0)
##   htmlwidgets            1.6.4     2023-12-06 [2] CRAN (R 4.5.0)
##   httpuv                 1.6.15    2024-03-26 [2] CRAN (R 4.5.0)
##   httr                   1.4.7     2023-08-15 [2] CRAN (R 4.5.0)
##   httr2                  1.1.0     2025-01-18 [2] CRAN (R 4.5.0)
##   igraph                 2.1.4     2025-01-23 [2] CRAN (R 4.5.0)
##   ini                    0.3.1     2018-05-20 [2] CRAN (R 4.5.0)
##   IRanges                2.41.2    2024-12-03 [2] Bioconductor 3.21 (R 4.5.0)
##   irlba                  2.3.5.1   2022-10-03 [2] CRAN (R 4.5.0)
##   isoband                0.2.7     2022-12-20 [2] CRAN (R 4.5.0)
##   iterators              1.0.14    2022-02-05 [2] CRAN (R 4.5.0)
##   jquerylib              0.1.4     2021-04-26 [2] CRAN (R 4.5.0)
##   jsonlite               1.8.9     2024-09-20 [2] CRAN (R 4.5.0)
##   KEGGREST               1.47.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   KernSmooth             2.23-26   2025-01-01 [3] CRAN (R 4.5.0)
##   knitr                  1.49      2024-11-08 [2] CRAN (R 4.5.0)
##   labeling               0.4.3     2023-08-29 [2] CRAN (R 4.5.0)
##   lambda.r               1.2.4     2019-09-18 [2] CRAN (R 4.5.0)
##   later                  1.4.1     2024-11-27 [2] CRAN (R 4.5.0)
##   lattice                0.22-6    2024-03-20 [3] CRAN (R 4.5.0)
##   lazyeval               0.2.2     2019-03-15 [2] CRAN (R 4.5.0)
##   lifecycle              1.0.4     2023-11-07 [2] CRAN (R 4.5.0)
##   limma                  3.63.3    2025-01-09 [2] Bioconductor 3.21 (R 4.5.0)
##   littler                0.3.20    2024-03-23 [2] CRAN (R 4.5.0)
##   lobstr                 1.1.2     2022-06-22 [2] CRAN (R 4.5.0)
##   locfit                 1.5-9.11  2025-02-03 [2] CRAN (R 4.5.0)
##   magick                 2.8.5     2024-09-20 [2] CRAN (R 4.5.0)
##   magrittr               2.0.3     2022-03-30 [2] CRAN (R 4.5.0)
##   MASS                   7.3-64    2025-01-04 [3] CRAN (R 4.5.0)
##   Matrix                 1.7-2     2025-01-23 [3] CRAN (R 4.5.0)
##   MatrixGenerics         1.19.1    2025-01-09 [2] Bioconductor 3.21 (R 4.5.0)
##   matrixStats            1.5.0     2025-01-07 [2] CRAN (R 4.5.0)
##   memoise                2.0.1     2021-11-26 [2] CRAN (R 4.5.0)
##   metapod                1.15.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   methods              * 4.5.0     2025-01-29 [3] local
##   mgcv                   1.9-1     2023-12-21 [3] CRAN (R 4.5.0)
##   mime                   0.12      2021-09-28 [2] CRAN (R 4.5.0)
##   miniUI                 0.1.1.1   2018-05-18 [2] CRAN (R 4.5.0)
##   munsell                0.5.1     2024-04-01 [2] CRAN (R 4.5.0)
##   nlme                   3.1-167   2025-01-27 [3] CRAN (R 4.5.0)
##   nnet                   7.3-20    2025-01-01 [3] CRAN (R 4.5.0)
##   nnSVG                  1.11.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   openssl                2.3.2     2025-02-03 [2] CRAN (R 4.5.0)
##   OSTA                   0.99.0    2025-02-07 [1] Bioconductor
##   paletteer              1.6.0     2024-01-21 [2] CRAN (R 4.5.0)
##   parallel               4.5.0     2025-01-29 [3] local
##   pbapply                1.7-2     2023-06-27 [2] CRAN (R 4.5.0)
##   pheatmap               1.0.12    2019-01-04 [2] CRAN (R 4.5.0)
##   pillar                 1.10.1    2025-01-07 [2] CRAN (R 4.5.0)
##   pkgbuild               1.4.6     2025-01-16 [2] CRAN (R 4.5.0)
##   pkgconfig              2.0.3     2019-09-22 [2] CRAN (R 4.5.0)
##   pkgdown                2.1.1     2024-09-17 [2] CRAN (R 4.5.0)
##   pkgload                1.4.0     2024-06-28 [2] CRAN (R 4.5.0)
##   plogr                  0.2.0     2018-03-25 [2] CRAN (R 4.5.0)
##   plotly                 4.10.4    2024-01-13 [2] CRAN (R 4.5.0)
##   png                    0.1-8     2022-11-29 [2] CRAN (R 4.5.0)
##   praise                 1.0.0     2015-08-11 [2] CRAN (R 4.5.0)
##   preprocessCore         1.69.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   prettyunits            1.2.0     2023-09-24 [2] CRAN (R 4.5.0)
##   prismatic              1.1.2     2024-04-10 [2] CRAN (R 4.5.0)
##   processx               3.8.5     2025-01-08 [2] CRAN (R 4.5.0)
##   profvis                0.4.0     2024-09-20 [2] CRAN (R 4.5.0)
##   promises               1.3.2     2024-11-28 [2] CRAN (R 4.5.0)
##   ps                     1.8.1     2024-10-28 [2] CRAN (R 4.5.0)
##   purrr                  1.0.4     2025-02-05 [2] CRAN (R 4.5.0)
##   R6                     2.5.1     2021-08-19 [2] CRAN (R 4.5.0)
##   ragg                   1.3.3     2024-09-11 [2] CRAN (R 4.5.0)
##   RANN                   2.6.2     2024-08-25 [2] CRAN (R 4.5.0)
##   rappdirs               0.3.3     2021-01-31 [2] CRAN (R 4.5.0)
##   rcmdcheck              1.4.0     2021-09-27 [2] CRAN (R 4.5.0)
##   RColorBrewer           1.1-3     2022-04-03 [2] CRAN (R 4.5.0)
##   Rcpp                   1.0.14    2025-01-12 [2] CRAN (R 4.5.0)
##   RcppAnnoy              0.0.22    2024-01-23 [2] CRAN (R 4.5.0)
##   RcppArmadillo          14.2.3-1  2025-02-06 [2] CRAN (R 4.5.0)
##   RcppEigen              0.3.4.0.2 2024-08-24 [2] CRAN (R 4.5.0)
##   RcppML                 0.3.7     2021-09-21 [2] CRAN (R 4.5.0)
##   RcppProgress           0.4.2     2020-02-06 [2] CRAN (R 4.5.0)
##   RCurl                  1.98-1.16 2024-07-11 [2] CRAN (R 4.5.0)
##   rdist                  0.0.5     2020-05-04 [2] CRAN (R 4.5.0)
##   rematch2               2.1.2     2020-05-01 [2] CRAN (R 4.5.0)
##   remotes                2.5.0     2024-03-17 [2] CRAN (R 4.5.0)
##   restfulr               0.0.15    2022-06-16 [2] CRAN (R 4.5.0)
##   Rhtslib                3.3.1     2024-11-27 [2] Bioconductor 3.21 (R 4.5.0)
##   rjson                  0.2.23    2024-09-16 [2] CRAN (R 4.5.0)
##   rlang                  1.1.5     2025-01-17 [2] CRAN (R 4.5.0)
##   rmarkdown              2.29      2024-11-04 [2] CRAN (R 4.5.0)
##   roxygen2               7.3.2     2024-06-28 [2] CRAN (R 4.5.0)
##   rpart                  4.1.24    2025-01-07 [3] CRAN (R 4.5.0)
##   rprojroot              2.0.4     2023-11-05 [2] CRAN (R 4.5.0)
##   Rsamtools              2.23.1    2024-11-27 [2] Bioconductor 3.21 (R 4.5.0)
##   RSpectra               0.16-2    2024-07-18 [2] CRAN (R 4.5.0)
##   RSQLite                2.3.9     2024-12-03 [2] CRAN (R 4.5.0)
##   rstudioapi             0.17.1    2024-10-22 [2] CRAN (R 4.5.0)
##   rsvd                   1.0.5     2021-04-16 [2] CRAN (R 4.5.0)
##   rtracklayer            1.67.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   Rtsne                  0.17      2023-12-07 [2] CRAN (R 4.5.0)
##   rversions              2.1.2     2022-08-31 [2] CRAN (R 4.5.0)
##   S4Arrays               1.7.2     2025-02-03 [2] Bioconductor 3.21 (R 4.5.0)
##   S4Vectors              0.45.2    2024-11-15 [2] Bioconductor 3.21 (R 4.5.0)
##   sass                   0.4.9     2024-03-15 [2] CRAN (R 4.5.0)
##   ScaledMatrix           1.15.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   scales                 1.3.0     2023-11-28 [2] CRAN (R 4.5.0)
##   scater                 1.35.1    2025-02-02 [2] Bioconductor 3.21 (R 4.5.0)
##   scran                  1.35.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   scuttle                1.17.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   sessioninfo            1.2.3     2025-02-05 [2] CRAN (R 4.5.0)
##   shape                  1.4.6.1   2024-02-23 [2] CRAN (R 4.5.0)
##   shiny                  1.10.0    2024-12-14 [2] CRAN (R 4.5.0)
##   shinyWidgets           0.8.7     2024-09-23 [2] CRAN (R 4.5.0)
##   SingleCellExperiment   1.29.1    2024-11-10 [2] Bioconductor 3.21 (R 4.5.0)
##   sitmo                  2.0.2     2021-10-13 [2] CRAN (R 4.5.0)
##   snow                   0.4-4     2021-10-27 [2] CRAN (R 4.5.0)
##   sourcetools            0.1.7-1   2023-02-01 [2] CRAN (R 4.5.0)
##   SparseArray            1.7.5     2025-02-03 [2] Bioconductor 3.21 (R 4.5.0)
##   spatial                7.3-18    2025-01-01 [3] CRAN (R 4.5.0)
##   SpatialExperiment      1.17.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   spatialLIBD            1.19.6    2024-12-17 [2] Bioconductor 3.21 (R 4.5.0)
##   splines                4.5.0     2025-01-29 [3] local
##   statmod                1.5.0     2023-01-06 [2] CRAN (R 4.5.0)
##   stats                * 4.5.0     2025-01-29 [3] local
##   stats4                 4.5.0     2025-01-29 [3] local
##   STexampleData          1.15.0    2024-11-05 [2] Bioconductor 3.21 (R 4.5.0)
##   stringi                1.8.4     2024-05-06 [2] CRAN (R 4.5.0)
##   stringr                1.5.1     2023-11-14 [2] CRAN (R 4.5.0)
##   SummarizedExperiment   1.37.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##   survival               3.8-3     2024-12-17 [3] CRAN (R 4.5.0)
##   sys                    3.4.3     2024-10-04 [2] CRAN (R 4.5.0)
##   systemfonts            1.2.1     2025-01-20 [2] CRAN (R 4.5.0)
##   tcltk                  4.5.0     2025-01-29 [3] local
##   testthat               3.2.3     2025-01-13 [2] CRAN (R 4.5.0)
##   textshaping            1.0.0     2025-01-20 [2] CRAN (R 4.5.0)
##   tibble                 3.2.1     2023-03-20 [2] CRAN (R 4.5.0)
##   tidyr                  1.3.1     2024-01-24 [2] CRAN (R 4.5.0)
##   tidyselect             1.2.1     2024-03-11 [2] CRAN (R 4.5.0)
##   tinytex                0.54      2024-11-01 [2] CRAN (R 4.5.0)
##   tools                  4.5.0     2025-01-29 [3] local
##   UCSC.utils             1.3.1     2025-01-15 [2] Bioconductor 3.21 (R 4.5.0)
##   urlchecker             1.0.1     2021-11-30 [2] CRAN (R 4.5.0)
##   usethis                3.1.0     2024-11-26 [2] CRAN (R 4.5.0)
##   utf8                   1.2.4     2023-10-22 [2] CRAN (R 4.5.0)
##   utils                * 4.5.0     2025-01-29 [3] local
##   uwot                   0.2.2     2024-04-21 [2] CRAN (R 4.5.0)
##   vctrs                  0.6.5     2023-12-01 [2] CRAN (R 4.5.0)
##   vipor                  0.4.7     2023-12-18 [2] CRAN (R 4.5.0)
##   viridis                0.6.5     2024-01-29 [2] CRAN (R 4.5.0)
##   viridisLite            0.4.2     2023-05-02 [2] CRAN (R 4.5.0)
##   waldo                  0.6.1     2024-11-07 [2] CRAN (R 4.5.0)
##   whisker                0.4.1     2022-12-05 [2] CRAN (R 4.5.0)
##   withr                  3.0.2     2024-10-28 [2] CRAN (R 4.5.0)
##   xfun                   0.50      2025-01-07 [2] CRAN (R 4.5.0)
##   XML                    3.99-0.18 2025-01-01 [2] CRAN (R 4.5.0)
##   xml2                   1.3.6     2023-12-04 [2] CRAN (R 4.5.0)
##   xopen                  1.0.1     2024-04-25 [2] CRAN (R 4.5.0)
##   xtable                 1.8-4     2019-04-21 [2] CRAN (R 4.5.0)
##   XVector                0.47.2    2025-01-08 [2] Bioconductor 3.21 (R 4.5.0)
##   yaml                   2.3.10    2024-07-26 [2] CRAN (R 4.5.0)
##   zip                    2.3.1     2024-01-27 [2] CRAN (R 4.5.0)
##   zlibbioc               1.53.0    2024-10-29 [2] Bioconductor 3.21 (R 4.5.0)
##  
##   [1] /tmp/RtmprlVBON/Rinst8c6e9d2db3
##   [2] /usr/local/lib/R/site-library
##   [3] /usr/local/lib/R/library
##   * â”€â”€ Packages attached to the search path.
##  
##  â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</code></pre>
</div>
</div>
</div>
</div>
</div>


</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../index.html" class="pagination-link" aria-label="Welcome">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Welcome</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/bkg-introduction.html" class="pagination-link" aria-label="Introduction">
        <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>